---
output:
  word_document: default
  html_document: default
---
Differential Gene Expression Data of Early Stages of Intestine Regeneration in Holothuria glaberrima
===============================================================================

**Authors:** *Joshua Medina, David Quispe* | 2020

```{r, echo=FALSE, message=FALSE, cache=FALSE}
library(BiocStyle)
knitr::opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE)
```

# Overview

The sea cucumber _Holothuria glaberrima_ has great regenerative capability after internal organs evisceration. In this case, we are analyzing data from mesentery regenerative tissue of 1 and 3 days post evisceration (1- and 3-DPE) in order to further understand the genes that are involved in the early stages of regeneration.

***

# Process

Set the directory to work in and load required libraries.

```{r}

setwd("D:/usuarios/david/Escritorio/Tesis/results/DESeq_protocol/DESeq_newprotocol/GitDESeq/")

suppressMessages(library(DESeq2))
suppressMessages(library(gplots))
suppressMessages(library(RColorBrewer))
suppressMessages(library(EnhancedVolcano))
suppressMessages(library(tidyverse))
suppressMessages(library(ggplot2))
suppressMessages(library(dplyr))
suppressMessages(library(tidyverse))
suppressMessages(library(RColorBrewer))
suppressMessages(library(calibrate))
suppressMessages(library(dichromat))
suppressMessages(library(IDPmisc))
suppressMessages(library(limma))
```


**Loading the data.**

* The variable *ddata* contains a csv file indicating the files to be used and to which group they correspond (i.e experimental or control). In our case this is going to consist of three different levels normal, 1 day post-evisceration (1DPE), and 3 days post-evisceration (3DPE).

```{r}
ddata <- read.table("data/pdata.txt", header = TRUE)
```




Generating the data set with respect to the counts. Here the count files generated with salmon are going to be used for the differential gene expression quantification.

```{r}
ddsSTAR <- DESeqDataSetFromHTSeqCount(sampleTable = ddata, directory = "counts_global", design = ~ condition)

colData(ddsSTAR)$condition <- factor(colData(ddsSTAR)$condition, levels=levels(ddata$condition))

#set a factor variable with six levels (conditions of the study). 
#Totally needed another option is to use relevel.

```




Filter transcripts with low count number. This works to eliminate noise from the different samples that have heterogeneous count number from artifact transcripts or samples with very low counts that can be discriminated.

```{r}
keep <- rowSums(counts(ddsSTAR)) >= 10
ddsSTAR <- ddsSTAR[keep,]
```




Estimating the size factors using _"median ratio method"_, dispersions (gene-wise, mean-dispersion relationship, final dispersion estimates). Generating the counts table with _counts_ function.

```{r}

dds <- estimateSizeFactors(ddsSTAR)
dds <- estimateDispersions(dds)
dds <- nbinomWaldTest(dds)
counts_table <- counts(dds, normalized=TRUE)

```




Results of contrasts between conditions (1DPE vs Normal; 3DPE vs Normal). __Note: Normal tissue is referred as non-regenerated mesentery tissue of the sea cucumber.__

```{r}
res_1DPE <- results(dds, contrast = c("condition", "1DPE", "Control"))
res_1DPE <- res_1DPE[order(res_1DPE$padj),]

res_3DPE <- results(dds, contrast = c("condition", "3DPE", "Control"))
res_3DPE <- res_3DPE[order(res_3DPE$padj),]

```




Assigning names to the transcripts
* You have to download the gene2pubmed table from ftp://ftp.ncbi.nih.gov/gene/DATA/

```{r}
File_1DPE<-as.data.frame(res_1DPE)
File_1DPE<-tibble::rownames_to_column(File_1DPE, "ID")
File_3DPE<-as.data.frame(res_3DPE)
File_3DPE<-tibble::rownames_to_column(File_3DPE, "ID")

#The GeneNames is the table generated with the annotation to identify the names of the annotated transcripts while the Final_NameMap links the IDs from the assembly with the IDs from the annotation
GeneNames<-read.csv("data/GeneNames.csv")
load("data/Final_NameMap.rda")

#The gene2pubmed has the pubmed IDs that enables the association with the SPU IDs from the reference_table. The SPU IDs will be necessary for the Gene Ontology analysis.
pubmedID<- read.delim("data/gene2pubmed", sep = "\t")
SPU_pmidIDs<-read.delim("data/reference_table.txt", header = FALSE, sep = "\t")

#Joining tables
CombinedTable_pmid<- GeneNames  %>% left_join(pubmedID, by= c("GeneID"="GeneID"))
Combined_Names<-CombinedTable_pmid %>% left_join(SPU_pmidIDs,by=c("PubMed_ID"="V2"))
tableCombined1DPE<-File_1DPE %>% left_join(Final_NameMap, by=c("ID"="Old_Name"))
tableCombined3DPE<-File_3DPE %>% left_join(Final_NameMap, by=c("ID"="Old_Name"))
Final1DPE<-tableCombined1DPE %>% left_join(Combined_Names, by=c("value"="Transcript"))
Final3DPE<-tableCombined3DPE %>% left_join(Combined_Names, by=c("value"="Transcript"))

#Eliminate duplicated produced from the numerous matches to avoid biased results
Final1DPE_dis<-distinct(Final1DPE, Final1DPE$ID, .keep_all = TRUE)
Final3DPE_dis<-distinct(Final3DPE, Final3DPE$ID, .keep_all = TRUE)
```




# Visualization of the results.

***

Plotting dispersion estimates.

```{r, fig.align = 'center'}
DESeq2::plotDispEsts(dds,main="Dispersion Plot")
```




Comparison of Raw Data vs normalized counts.

```{r, fig.align = 'center'}

conds <- as.vector(ddata$condition)
condcols <- brewer.pal(n = length(unique(conds)), name="Dark2") 
par(mfrow=c(1,2))
barplot(colSums(counts(dds, normalized=F)), col = condcols[as.factor(conds)], las = 2, cex.names=0.6, main="Raw Counts")
barplot(colSums(counts(dds, normalized=T)), col=condcols[as.factor(conds)], las=2, cex.names = 0.6, main = "Normalized Counts")

```




PCA Plot

```{r, fig.align = 'center'}

vsd <- varianceStabilizingTransformation(dds)
DESeq2::plotPCA(vsd)+theme_bw()

```




Volcano plot of Normal vs 1DPE

```{r}
with(res_1DPE, plot(log2FoldChange, -log10(pvalue), pch=20, main="Time0 vs 1DPE", xlim=c(-30,30)))

with(subset(res_1DPE, padj<=0.05 & log2FoldChange>=2), points(log2FoldChange, -log10(pvalue), pch=20, col="red"))
with(subset(res_1DPE, padj<=0.05 & log2FoldChange<=-2), points(log2FoldChange, -log10(pvalue), pch=20, col="blue"))

abline(h=-log10(0.05),lty=3)
abline(v=-1,lty=3)
abline(v=1,lty=3)
```




Volcano plot of Normal vs 3DPE

```{r}
with(res_3DPE, plot(log2FoldChange, -log10(pvalue), pch=20, main="Time0 vs 3DPE", xlim=c(-30,30)))

with(subset(res_3DPE, padj<=0.05 & log2FoldChange>=2), points(log2FoldChange, -log10(pvalue), pch=20, col="red"))
with(subset(res_3DPE, padj<=0.05 & log2FoldChange<=-2), points(log2FoldChange, -log10(pvalue), pch=20, col="blue"))

abline(h=-log10(0.05),lty=3)
abline(v=-1,lty=3)
abline(v=1,lty=3)
```




Gene Ontology


```{r}
#Load GO table from echinobase
GO_ids<-read.delim("data/go_table.txt", sep = "\t", colClasses = c("factor", "character","character","character","character"))

#Get packages
suppressMessages(library(GO.db))
suppressMessages(library(GSEABase))
suppressMessages(library(GOstats))
suppressMessages(library(GOplot))

```



1 DPE
```{r}
#Table combination with SPU IDs to obatin GO IDs
GO_1DPE_all<-Final1DPE_dis %>% left_join(GO_ids, by = c("V1"="SPU_ID"))

#Eliminate rows with NA from table to avoid errors in the subsequent commands
GO_1DPE_all<-na.omit(GO_1DPE_all)

#Generate a table with the necessary columns
GO_1DPE<-GO_1DPE_all[,c(22,23,14,3,8)]

#Add the GO: to the GO IDs to be recognized in the analysis
GO_1DPE$GO_ID <- sub("^", "GO:", GO_1DPE$GO_ID)

#Generate a data frame for the GO analysis with three columns
goframeData = data.frame(GO_1DPE$GO_ID, GO_1DPE$Evidence, GO_1DPE$value)

goFrame=GOFrame(goframeData, organism="Holothuria glaberrima")

goAllFrame=GOAllFrame(goFrame)

gsc <- GeneSetCollection(goAllFrame, setType = GOCollection())

universe <- GO_1DPE_all$value

genes <- GO_1DPE_all$value[GO_1DPE_all$padj < 0.05]

#For the GO analysis we are only going to test the over expressed GO terms from the data

#Parameters for molecular function (MF) GO
params_mf <- GSEAGOHyperGParams(name="My Custom GSEA based annot Params_MF",
                                geneSetCollection=gsc,
                                geneIds=genes,
                                universeGeneIds=universe,
                                ontology="MF",
                                pvalueCutoff=0.05,
                                conditional=FALSE,
                                testDirection="over")

Over_MF <- hyperGTest(params_mf)

#Parameters for cellular component (CC) GO
params_cc <- GSEAGOHyperGParams(name="My Custom GSEA based annot Params_CC",
                                geneSetCollection=gsc,
                                geneIds=genes,
                                universeGeneIds=universe,
                                ontology="CC",
                                pvalueCutoff=0.05,
                                conditional=FALSE,
                                testDirection="over")

Over_CC <- hyperGTest(params_cc)

#Parameters for biological process (BP) GO
params_bp <- GSEAGOHyperGParams(name="My Custom GSEA based annot Params_BP",
                                geneSetCollection=gsc,
                                geneIds=genes,
                                universeGeneIds=universe,
                                ontology="BP",
                                pvalueCutoff=0.05,
                                conditional=FALSE,
                                testDirection="over")

Over_BP <- hyperGTest(params_bp)


#Summary of the GO results
Over_BP_df<-summary(Over_BP)
Over_CC_df<-summary(Over_CC)
Over_MF_df<-summary(Over_MF)

#Assign of the category of each GO ID to another column
Over_BP_df<-cbind(Over_BP_df, category="BP")
Over_CC_df<-cbind(Over_CC_df, category="CC")
Over_MF_df<-cbind(Over_MF_df, category="MF")

#Change of name of the ID in each data frame summary
names(Over_BP_df)[names(Over_BP_df) == "GOBPID"] <- "ID"
names(Over_CC_df)[names(Over_CC_df) == "GOCCID"] <- "ID"
names(Over_MF_df)[names(Over_MF_df) == "GOMFID"] <- "ID"

#Combination of results in a single data frame
GO_1DPE_df<-rbind(Over_BP_df,Over_CC_df,Over_MF_df)

#Function to determine the z-score from the Odds ratio and the p-value 
zscore<-function(a,b){
  z=sign(log(a))*abs(qnorm(b/2))
  return(z)
}

#Apply the function to all the rows and assign the result to another column
GO_1DPE_df$Z_score<-mapply(zscore, GO_1DPE_df$OddsRatio, GO_1DPE_df$Pvalue)

#Join GO results to data frame with gene names and expression
GO_Finaldf<- GO_1DPE %>% inner_join(GO_1DPE_df, by=c("GO_ID"="ID"))

#Change colnames 
colnames(GO_Finaldf)<-c("ID","Evidence","genes","logFC","value","adj_pval","OddsRation", "ExpCount","count","Size","term","category","zscore")

#Extract columns for data frame used in GOplot package
circ<-GO_Finaldf[,c(12,1,11,9,3,4,6,13)]

#Choose GO IDs from the results obtained and store them in a variable
IDs_GO1DPE<-distinct(circ, circ$ID)
IDs_GO1DPE<-IDs_GO1DPE$`circ$ID`

#Plot GO results for over expresed GO terms
dev.new(width=10, height=10)
GOCircle(circ, nsub = IDs_GO1DPE, table.legend = TRUE, label.size = 3.5)

```




3 DPE
```{r}
#Table combination with SPU IDs to obatin GO IDs
GO_3DPE_all<-Final3DPE_dis %>% left_join(GO_ids, by = c("V1"="SPU_ID"))

#Eliminate rows with NA from table to avoid errors in the subsequent commands
GO_3DPE_all<-na.omit(GO_3DPE_all)

#Generate a table with the necessary columns
GO_3DPE<-GO_3DPE_all[,c(22,23,14,3,8)]

#Add the GO: to the GO IDs to be recognized in the analysis
GO_3DPE$GO_ID <- sub("^", "GO:", GO_3DPE$GO_ID)

#Generate a data frame for the GO analysis with three columns
goframeData_3 = data.frame(GO_3DPE$GO_ID, GO_3DPE$Evidence, GO_3DPE$value)

goFrame_3=GOFrame(goframeData_3, organism="Holothuria glaberrima")

goAllFrame_3=GOAllFrame(goFrame_3)

gsc_3 <- GeneSetCollection(goAllFrame_3, setType = GOCollection())

universe_3 <- GO_3DPE_all$value

genes_3 <- GO_3DPE_all$value[GO_3DPE_all$padj < 0.05]

#Parameters for molecular function (MF) GO
params_mf_3 <- GSEAGOHyperGParams(name="My Custom GSEA based annot Params_MF",
                                geneSetCollection=gsc_3,
                                geneIds=genes_3,
                                universeGeneIds=universe_3,
                                ontology="MF",
                                pvalueCutoff=0.05,
                                conditional=FALSE,
                                testDirection="over")

Over_MF_3 <- hyperGTest(params_mf_3)

#Parameters for cellular component (CC) GO
params_cc_3 <- GSEAGOHyperGParams(name="My Custom GSEA based annot Params_CC",
                                geneSetCollection=gsc_3,
                                geneIds=genes_3,
                                universeGeneIds=universe_3,
                                ontology="CC",
                                pvalueCutoff=0.05,
                                conditional=FALSE,
                                testDirection="over")

Over_CC_3 <- hyperGTest(params_cc_3)

#Parameters for biological process (BP) GO
params_bp_3 <- GSEAGOHyperGParams(name="My Custom GSEA based annot Params_BP",
                                geneSetCollection=gsc_3,
                                geneIds=genes_3,
                                universeGeneIds=universe_3,
                                ontology="BP",
                                pvalueCutoff=0.05,
                                conditional=FALSE,
                                testDirection="over")

Over_BP_3 <- hyperGTest(params_bp_3)


#Summary of the GO results
Over_BP_df3<-summary(Over_BP_3)
Over_CC_df3<-summary(Over_CC_3)
Over_MF_df3<-summary(Over_MF_3)

#Assign of the category of each GO ID to another column
Over_BP_df3<-cbind(Over_BP_df3, category="BP")
Over_CC_df3<-cbind(Over_CC_df3, category="CC")
Over_MF_df3<-cbind(Over_MF_df3, category="MF")

#Change of name of the ID in each data frame summary
names(Over_BP_df3)[names(Over_BP_df3) == "GOBPID"] <- "ID"
names(Over_CC_df3)[names(Over_CC_df3) == "GOCCID"] <- "ID"
names(Over_MF_df3)[names(Over_MF_df3) == "GOMFID"] <- "ID"

#Combination of results in a single data frame
GO_3DPE_df<-rbind(Over_BP_df3,Over_CC_df3,Over_MF_df3)

#Apply the function (previously made) to all the rows and assign the result to another column
GO_3DPE_df$Z_score<-mapply(zscore, GO_3DPE_df$OddsRatio, GO_3DPE_df$Pvalue)

#Join GO results to data frame with gene names and expression
GO_Finaldf3<- GO_3DPE %>% inner_join(GO_3DPE_df, by=c("GO_ID"="ID"))

#Change colnames 
colnames(GO_Finaldf3)<-c("ID","Evidence","genes","logFC","value","adj_pval","OddsRation", "ExpCount","count","Size","term","category","zscore")

#Extract columns for data frame used in GOplot package
circ3<-GO_Finaldf3[,c(12,1,11,9,3,4,6,13)]

#Choose GO IDs from the results obtained and store them in a variable
IDs_GO3DPE<-distinct(circ3, circ3$ID)
IDs_GO3DPE<-IDs_GO3DPE$`circ3$ID`

#Plot GO results for over expresed GO terms
GOCircle(circ3, nsub = IDs_GO3DPE, table.legend = TRUE, label.size = 4)

```



Pathway Enrichment

1 DPE
```{r}
#Filtering IDs with <0.05 of adjusted p-value
keggIDs_1DPE <- Final1DPE_dis$GeneID[Final1DPE_dis$padj < 0.05]

#Kegga command from limma package
kegg_1DPE<-kegga(keggIDs_1DPE, species.KEGG = "spu")

#Calculate GeneRation from DE and N
kegg_1DPE$GeneRatio<-kegg_1DPE$DE/kegg_1DPE$N
colnames(kegg_1DPE)<-c("Pathway","N","Count","Pvalue","GeneRatio")

#Using ggplot to visualize results
ggplot(filter(kegg_1DPE, kegg_1DPE$Pvalue<0.01)) +
    geom_point(aes(x = GeneRatio, y = reorder(Pathway, GeneRatio),
               size = Count, color = Pvalue))+ylab("")+theme_bw()+scale_color_gradient2(low = 'blue', mid = 'red')+scale_size_area()
```

3 DPE
```{r}
#Filtering IDs with <0.05 of adjusted p-value
keggIDs_3DPE <- Final3DPE_dis$GeneID[Final3DPE_dis$padj < 0.05]

#Kegga command from limma package
kegg_3DPE<-kegga(keggIDs_3DPE, species.KEGG = "spu")

#Calculate GeneRation from DE and N
kegg_3DPE$GeneRatio<-kegg_3DPE$DE/kegg_3DPE$N
colnames(kegg_3DPE)<-c("Pathway","N","Count","Pvalue","GeneRatio")

#Using ggplot to visualize results
ggplot(filter(kegg_3DPE, kegg_3DPE$Pvalue<0.01)) +
    geom_point(aes(x = GeneRatio, y = reorder(Pathway, GeneRatio),
               size = Count, color = Pvalue))+ylab("")+theme_bw()+scale_color_gradient2(low = 'blue', mid = 'red')+scale_size_area()
```

